[#1530] Create new controls UI for video playback

kroucis opened the pull request:
Addresses #1530

Create a new controls UI layer for the DamusVideoPlayer, providing play/pause, volume control, content scale, playback rate controls, as well as a seek/timecode bar.

<img width="594" alt="Screenshot 2023-09-05 at 11 42 46" src="https://github.com/damus-io/damus/assets/1031501/de89196e-0951-4a59-8430-99a8cd166e60">
<img width="586" alt="Screenshot 2023-09-05 at 11 42 55" src="https://github.com/damus-io/damus/assets/1031501/4a947870-640d-4c7e-89f7-1eab30326aa3">

## Features
- Scrub on the horizontal seek bar to change the playback time
- Play/Pause button
- Mute/Unmute button and volume vertical slider
- Select playback rate from 0.5x up to 2.0x
- Tap video to hide/show controls
- Swap between aspect fill and aspect fit
- Player has option to hide controls on startup
- Player's video tap behavior can be customized
## Fixes
- The VideoPlayer has been reworked from frame-by-frame updating to an event-driven model, significantly improving performance and understandability.
- The VideoPlayer and DamusVideoPlayer had a number of leaks that were preventing resource cleanup and causing a lot of conflicting video playback behavior.
## TODOs
- Animate the volume slider in/out
- Animate the entire control layer in/out
- Create a timer to automatically hide the controls after some time of inactivity
- Color/Theme unification
## Known Issues
- This still suffers from [#1386] as that is a navigation/data model architecture issue, not specific to the video player.
## Notes
- !! This feature *DOES NOT* INCLUDE A FIX FOR #1386 ! That fix will be provided in a future PR. !!
## Support
Bounty: 100k satoshis(BTC) -> 0.001 BTC
Payout Address: `bc1q5sxugecyhq5stxn4d79p4908hdr7ljcn35uyk7`
<img width="341" alt="btc001" src="https://github.com/damus-io/damus/assets/1031501/54552b4f-ad4e-4842-b1a2-2ac44bf47ad7">

kroucis commented:
The iOS 16.2 test failed because of a missing method that is gated by a check for iOS 16.4? What am I missing here? That seems like a test suite bug...

jb55 commented:
On Tue, Sep 05, 2023 at 09:52:45AM -0700, Kyle Roucis wrote:
>The iOS 16.2 test failed because of a missing method that is gated by a check for iOS 16.4? What am I missing here? That seems like a test suite bug...

test suite is very buggy

kroucis commented:
Well I tested these changes on a 16.0 simulator instance. Compiled and ran just fine.

jb55 commented:
Reviewed-by: William Casarin ***@***.***>

On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote:
>Closes: https://github.com/damus-io/damus/pull/1537
>---
> damus/Views/Video/VideoPlayer.swift | 112 ++++++++++++++++++++--------
> 1 file changed, 82 insertions(+), 30 deletions(-)
>
>diff --git a/damus/Views/Video/VideoPlayer.swift b/damus/Views/Video/VideoPlayer.swift
>index 2a0a04d49..53f25323e 100644
>--- a/damus/Views/Video/VideoPlayer.swift
>+++ b/damus/Views/Video/VideoPlayer.swift
>@@ -15,6 +15,7 @@ import Foundation
> //
>
> import AVFoundation
>+import Combine
> import GSPlayer
> import SwiftUI
>
>@@ -47,8 +48,13 @@ public class VideoPlayerModel: ObservableObject {
>     @Published var size: CGSize? = nil
>     @Published var has_audio: Bool? = nil
>     @Published var contentMode: UIView.ContentMode = .scaleAspectFill
>+    @Published var currentTime: Double = 0.0
>+    @Published var playbackRate: Float = 1.0
>+    @Published var volume: Float = 1.0
>+    // Split out a deliberate stream for *setting* currentTime manually, that way system-driven updates don't cause infinite loops and strange fighting behavior.
>+    let currentTimeSubject = PassthroughSubject<Double, Never>()
>
>-    fileprivate var time: CMTime?
>+    var totalDuration: Double = 0.0
>
>     var handlers: [VideoHandler] = []
>
>@@ -72,20 +78,36 @@ public class VideoPlayerModel: ObservableObject {
>     }
>
>     /// Whether the video will be automatically replayed until the end of the video playback.
>-    func autoReplay(_ value: Bool) -> Self {
>+    func set(autoReplay value: Bool) -> Self {
>         autoReplay = value
>         return self
>     }
>
>     /// Whether the video is muted, only for this instance.
>-    func mute(_ value: Bool) -> Self {
>+    func set(muted value: Bool) -> Self {
>         muted = value
>         return self
>     }
>
>+    func set(volume value: Float) -> Self {
>+        volume = value
>+        return self
>+    }
>+
>+    func set(playbackRate value: Float) -> Self {
>+        self.playbackRate = value
>+        return self
>+    }
>+
>+    func set(seekSeconds: Double) -> Self {
>+        self.currentTime = seekSeconds
>+        self.currentTimeSubject.send(self.currentTime)
>+        return self
>+    }
>+
>     /// A string defining how the video is displayed within an AVPlayerLayer bounds rect.
>     /// scaleAspectFill -> resizeAspectFill, scaleAspectFit -> resizeAspect, other -> resize
>-    func contentMode(_ value: UIView.ContentMode) -> Self {
>+    func set(contentMode value: UIView.ContentMode) -> Self {
>         contentMode = value
>         return self
>     }
>@@ -166,16 +188,22 @@ public extension VideoPlayer {
>     }
> }
>
>-func get_video_size(player: AVPlayer) async -> CGSize? {
>-    let res = Task.detached(priority: .background) {
>-        return player.currentImage?.size
>+fileprivate extension AVPlayer {
>+    var videoSize: CGSize? {
>+        get async {
>+            let res = Task.detached(priority: .background) {
>+                return self.currentImage?.size
>+            }
>+            return await res.value
>+        }
>+    }
>+
>+    var hasAudio: Bool {
>+        get async {
>+            let tracks = try? await self.currentItem?.asset.load(.tracks)
>+            return tracks?.filter({ t in t.mediaType == .audio }).first != nil
>+        }
>     }
>-    return await res.value
>-}
>-
>-func video_has_audio(player: AVPlayer) async -> Bool {
>-    let tracks = try? await player.currentItem?.asset.load(.tracks)
>-    return tracks?.filter({ t in t.mediaType == .audio }).first != nil
> }
>
> @available(iOS 13, *)
>@@ -210,15 +238,15 @@ extension VideoPlayer: UIViewRepresentable {
>         }
>
>         uiView.stateDidChanged = { [unowned uiView] _ in
>-            let state: VideoState = uiView.convertState()
>+            let state: VideoState = uiView.videoState
>
>-            if case .playing = state {
>+            if case .playing = uiView.videoState {
>                 context.coordinator.startObserver(uiView: uiView)
>
>                 if let player = uiView.player {
>                     Task {
>-                        let has_audio = await video_has_audio(player: player)
>-                        let size = await get_video_size(player: player)
>+                        let has_audio = await player.hasAudio
>+                        let size = await player.videoSize
>                         Task { @MainActor in
>                             if let size {
>                                 self.model.size = size
>@@ -262,11 +290,6 @@ extension VideoPlayer: UIViewRepresentable {
>
>         uiView.isMuted = model.muted
>         uiView.isAutoReplay = model.autoReplay
>-
>-        if let observerTime = context.coordinator.observerTime, let modelTime = model.time,
>-           modelTime != observerTime && modelTime.isValid && modelTime.isNumeric {
>-            uiView.seek(to: modelTime, completion: { _ in })
>-        }
>     }
>
>     public static func dismantleUIView(_ uiView: VideoPlayerView, coordinator: VideoPlayer.Coordinator) {
>@@ -277,8 +300,11 @@ extension VideoPlayer: UIViewRepresentable {
>         var videoPlayer: VideoPlayer
>         var observingURL: URL?
>         var observer: Any?
>-        var observerTime: CMTime?
>         var observerBuffer: Double?
>+        var videoModeSub: AnyCancellable?
>+        var playbackRateSub: AnyCancellable?
>+        var volumeSub: AnyCancellable?
>+        var currentTimeSub: AnyCancellable?
>
>         init(_ videoPlayer: VideoPlayer) {
>             self.videoPlayer = videoPlayer
>@@ -288,31 +314,58 @@ extension VideoPlayer: UIViewRepresentable {
>         func startObserver(uiView: VideoPlayerView) {
>             guard observer == nil else { return }
>
>+            self.videoModeSub = self.videoPlayer.model.$contentMode.sink { mode in
>+                uiView.contentMode = mode
>+            }
>+
>+            self.playbackRateSub = self.videoPlayer.model.$playbackRate.sink { rate in
>+                uiView.player?.rate = rate
>+            }
>+
>+            self.volumeSub = self.videoPlayer.model.$volume.sink { volume in
>+                uiView.player?.volume = volume
>+            }
>+
>+            self.currentTimeSub = self.videoPlayer.model.currentTimeSubject.sink { seconds in
>+                let currentTimeScale = uiView.player?.currentTime().timescale ?? 1
>+                let newTime = CMTimeMakeWithSeconds(seconds, preferredTimescale: currentTimeScale)
>+                let halfSecondBefore = CMTimeMakeWithSeconds(seconds - 0.5, preferredTimescale: currentTimeScale)
>+                let halfSecondAfter = CMTimeMakeWithSeconds(seconds + 0.5, preferredTimescale: currentTimeScale)
>+                uiView.player?.seek(to: newTime,
>+                                    toleranceBefore: halfSecondBefore,
>+                                    toleranceAfter: halfSecondAfter)
>+            }
>+
>             observer = uiView.addPeriodicTimeObserver(forInterval: .init(seconds: 0.25, preferredTimescale: 60)) { [weak self, unowned uiView] time in
>-                guard let `self` = self else { return }
>+                guard let self else { return }
>
>+                self.videoPlayer.model.totalDuration = uiView.player?.totalDuration ?? 0.0
>                 Task { @MainActor in
>-                    self.videoPlayer.model.time = time
>+                    self.videoPlayer.model.currentTime = CMTimeGetSeconds(time)
>                 }
>-                self.observerTime = time
>
>                 self.updateBuffer(uiView: uiView)
>             }
>         }
>
>         func stopObserver(uiView: VideoPlayerView) {
>-            guard let observer = observer else { return }
>+            guard let observer else { return }
>
>             uiView.removeTimeObserver(observer)
>
>             self.observer = nil
>+
>+            // TODO: Should this call `clean()`?
>         }
>
>         func clean() {
>             self.observingURL = nil
>             self.observer = nil
>-            self.observerTime = nil
>             self.observerBuffer = nil
>+            self.videoModeSub = nil
>+            self.playbackRateSub = nil
>+            self.volumeSub = nil
>+            self.currentTimeSub = nil
>         }
>
>         @MainActor
>@@ -334,8 +387,7 @@ extension VideoPlayer: UIViewRepresentable {
> }
>
> private extension VideoPlayerView {
>-
>-    func convertState() -> VideoState {
>+    var videoState: VideoState {
>         switch state {
>         case .none, .loading:
>             return .loading
>

bryanmontz commented:
Yeah it’ll definitely conflict. I’m removing GSPlayer and making a lot of simplifications in DamusVideoPlayer. If you want, you can finish out this task and merge it when you’re happy, then I can deal with the conflicts. I might be able to contribute a review tonight, but not sure.



On Sep 5, 2023, at 11:49 AM, William Casarin ***@***.***> wrote:


Hey Kyle & Bryan,

On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote:
>Addresses #1530
>
>Create a new controls UI layer for the DamusVideoPlayer, providing play/pause, volume control, content scale, playback rate controls, as well as a seek/timecode bar.
>
>&lt;img width=&quot;594&quot; alt=&quot;Screenshot 2023-09-05 at 11 42 46&quot; src=&quot;https://github.com/damus-io/damus/assets/1031501/de89196e-0951-4a59-8430-99a8cd166e60&quot;&gt;
>&lt;img width=&quot;586&quot; alt=&quot;Screenshot 2023-09-05 at 11 42 55&quot; src=&quot;https://github.com/damus-io/damus/assets/1031501/4a947870-640d-4c7e-89f7-1eab30326aa3&quot;&gt;

Definitely needs a design pass. not sure about the purple. Not a blocker though, we can fix this up after.

>Current features:
>[..]
>- The VideoPlayer has been reworked from frame-by-frame updating to an event-driven model, significantly improving performance and understandability.
>Known Issues:
>- This still suffers from [#1386] as that is a navigation/data model architecture issue, not specific to the video player.
>Note:
>- This feature DOES NOT INCLUDE A FIX FOR #1386 ! That fix will be provided in a future PR.

Does this conflict with the updates you're making Bryan?


Will

—
Reply to this email directly, view it on GitHub<https://github.com/damus-io/damus/pull/1537#issuecomment-1706967904>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AADM3OR55SHXF6UMCXVB7N3XY5JYRANCNFSM6AAAAAA4L77HZI>.
You are receiving this because your review was requested.Message ID: ***@***.***>

jb55 commented:
On Tue, Sep 05, 2023 at 09:55:55AM -0700, William Casarin wrote:
>On Tue, Sep 05, 2023 at 09:52:45AM -0700, Kyle Roucis wrote:
>>The iOS 16.2 test failed because of a missing method that is gated by a check for iOS 16.4? What am I missing here? That seems like a test suite bug...
>
>test suite is very buggy

I would still like to not break the CI though, we should try to
workaround this somehow.

kroucis commented:
> On Tue, Sep 05, 2023 at 09:55:55AM -0700, William Casarin wrote: On Tue, Sep 05, 2023 at 09:52:45AM -0700, Kyle Roucis wrote: >The iOS 16.2 test failed because of a missing method that is gated by a check for iOS 16.4? What am I missing here? That seems like a test suite bug... test suite is very buggy
> I would still like to not break the CI though, we should try to workaround this somehow.

A simple but unappealing fix would be to NOT use the popover call. This turns the popover into a sheet. Looks like crap, but is supported on all versions of SwiftUI

jb55 commented:
On Tue, Sep 05, 2023 at 10:18:50AM -0700, Bryan Montz wrote:
>Yeah it’ll definitely conflict. I’m removing GSPlayer and making a lot of simplifications in DamusVideoPlayer. If you want, you can finish out this task and merge it when you’re happy, then I can deal with the conflicts. I might be able to contribute a review tonight, but not sure.

As long as we can still remove GSPlayer after this change then I'm
happy. I prefer not to have dependencies that we don't need.

jb55 commented:
On Tue, Sep 05, 2023 at 01:42:00PM -0400, Grimless wrote:
>Closes: https://github.com/damus-io/damus/pull/1537
>---
> damus/Views/Video/DamusVideoPlayer.swift | 7 ++-----
> 1 file changed, 2 insertions(+), 5 deletions(-)
>
>diff --git a/damus/Views/Video/DamusVideoPlayer.swift b/damus/Views/Video/DamusVideoPlayer.swift
>index b6e92346f..1ae325347 100644
>--- a/damus/Views/Video/DamusVideoPlayer.swift
>+++ b/damus/Views/Video/DamusVideoPlayer.swift
>@@ -122,11 +122,8 @@ struct DamusVideoPlayer: View {
>             .popover(isPresented: $isOptionsPopoverPresented,
>                      attachmentAnchor: .point(.top),
>                      arrowEdge: .top) {
>-                if #available(iOS 16.4, macOS 13.3, *) {
>-                    OptionsMenu
>-                        .presentationCompactAdaptation(.popover)
>-                }

I think we should just fix the CI instead of dropping this if it looks better.

>-                else {
>+                VStack {
>+                    Text("Playback Speed").font(.title)
>                     OptionsMenu
>                 }
>             }

jb55 commented:
On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote:
>Current features:
>- Scrub on the horizontal seek bar to change the playback time
>- Play/Pause button
>- Mute/Unmute button and volume vertical slider
>- Select playback rate from 0.5x up to 2.0x
>- Tap video to hide/show controls
>- Player has option to hide controls on startup
>- Player&#39;s video tap behavior can be customized

It looks like the fullscreen button doesn't do anything? I don't see it
listed here so I assume that's not implemented?

kroucis commented:
> On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote: Current features: - Scrub on the horizontal seek bar to change the playback time - Play/Pause button - Mute/Unmute button and volume vertical slider - Select playback rate from 0.5x up to 2.0x - Tap video to hide/show controls - Player has option to hide controls on startup - Player&#39;s video tap behavior can be customized
> It looks like the fullscreen button doesn't do anything? I don't see it listed here so I assume that's not implemented?

No it functions: it swaps between the available functionality for content fitting: .aspectFill and .aspectFit. There does not seem to be "fullscreen" support with the current video player dependency. Maybe @bryanmontz 's upcoming changes can support more, but there was no clear way to implement "fullscreen", just different fitting ratios.

jb55 commented:
I'm hitting crashes, here one I just hit when testing this branch

Fatal error: Attempted to read an unowned reference but the object was already deallocated

Thread 1 Queue : com.apple.main-thread (serial)
#0	0x00000001d55c401c in __pthread_kill ()
#1	0x00000001f7ded680 in pthread_kill ()
#2	0x0000000197ea3bb0 in abort ()
#3	0x00000001897094d8 in swift::fatalErrorv(unsigned int, char const*, char*) ()
#4	0x00000001897094f8 in swift::fatalError(unsigned int, char const*, ...) ()
#5	0x00000001897096b4 in swift::swift_abortRetainUnowned(void const*) ()
#6	0x0000000189766018 in swift_unknownObjectUnownedLoadStrong ()
#7	0x00000001006cafdc in closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) at /Users/jb55/dev/github/damus-io/damus/damus/Views/Video/VideoPlayer.swift:345
#8	0x00000001006cfe1c in partial apply for closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) ()
#9	0x00000001006c7a9c in thunk for @escaping @callee_guaranteed @Sendable @async () -> () ()
#10	0x00000001006cff60 in partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> () ()
#11	0x000000010069dc54 in thunk for @escaping @callee_guaranteed @Sendable @async () -> ***@***.*** τ_0_0) ()
#12	0x000000010069ddb0 in partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> ***@***.*** τ_0_0) ()

VideoPlayer.swift:342
342   observer = uiView.addPeriodicTimeObserver(forInterval: .init(seconds: 0.25, preferredTimescale: 60)) { [weak self, unowned uiView] time in
343       guard let self else { return }
344       
345       Task { @MainActor in  // <--- !!!CRASH HERE !!!
346           self.videoPlayer.model.currentTime = uiView.currentDuration
347       }
348       
349       self.updateBuffer(uiView: uiView)
350   }

jb55 commented:
On Tue, Sep 05, 2023 at 11:50:21AM -0700, Kyle Roucis wrote:
>> On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote: Current features: - Scrub on the horizontal seek bar to change the playback time - Play/Pause button - Mute/Unmute button and volume vertical slider - Select playback rate from 0.5x up to 2.0x - Tap video to hide/show controls - Player has option to hide controls on startup - Player&#39;s video tap behavior can be customized
>> It looks like the fullscreen button doesn't do anything? I don't see it listed here so I assume that's not implemented?
>
>No it functions: it swaps between the available functionality for content fitting: .aspectFill and .aspectFit. There does not seem to be "fullscreen" support with the current video player dependency. Maybe @bryanmontz 's upcoming changes can support more, but there was no clear way to implement "fullscreen", just different fitting ratios.

ok we should remove that button then until we have fullscreen.

kroucis commented:
> On Tue, Sep 05, 2023 at 11:50:21AM -0700, Kyle Roucis wrote:
> >> On Mon, Sep 04, 2023 at 11:33:14PM -0400, Grimless wrote: Current features: - Scrub on the horizontal seek bar to change the playback time - Play/Pause button - Mute/Unmute button and volume vertical slider - Select playback rate from 0.5x up to 2.0x - Tap video to hide/show controls - Player has option to hide controls on startup - Player&#39;s video tap behavior can be customized
> >> It looks like the fullscreen button doesn't do anything? I don't see it listed here so I assume that's not implemented?
> >
> >No it functions: it swaps between the available functionality for content fitting: .aspectFill and .aspectFit. There does not seem to be "fullscreen" support with the current video player dependency. Maybe @bryanmontz 's upcoming changes can support more, but there was no clear way to implement "fullscreen", just different fitting ratios.
> 
> ok we should remove that button then until we have fullscreen.
> 

Oh! Uh are you sure? I still find the aspect change to be helpful even without fullscreen support.

kroucis commented:
> 
> I'm hitting crashes, here one I just hit when testing this branch
> 
> Fatal error: Attempted to read an unowned reference but the object was already deallocated
> 
> Thread 1 Queue : com.apple.main-thread (serial)
> #0	0x00000001d55c401c in __pthread_kill ()
> #1	0x00000001f7ded680 in pthread_kill ()
> #2	0x0000000197ea3bb0 in abort ()
> #3	0x00000001897094d8 in swift::fatalErrorv(unsigned int, char const*, char*) ()
> #4	0x00000001897094f8 in swift::fatalError(unsigned int, char const*, ...) ()
> #5	0x00000001897096b4 in swift::swift_abortRetainUnowned(void const*) ()
> #6	0x0000000189766018 in swift_unknownObjectUnownedLoadStrong ()
> #7	0x00000001006cafdc in closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) at /Users/jb55/dev/github/damus-io/damus/damus/Views/Video/VideoPlayer.swift:345
> #8	0x00000001006cfe1c in partial apply for closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) ()
> #9	0x00000001006c7a9c in thunk for @escaping @callee_guaranteed @Sendable @async () -> () ()
> #10	0x00000001006cff60 in partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> () ()
> #11	0x000000010069dc54 in thunk for @escaping @callee_guaranteed @Sendable @async () -> ***@***.*** τ_0_0) ()
> #12	0x000000010069ddb0 in partial apply for thunk for @escaping @callee_guaranteed @Sendable @async () -> ***@***.*** τ_0_0) ()
> 
> VideoPlayer.swift:342
> 342   observer = uiView.addPeriodicTimeObserver(forInterval: .init(seconds: 0.25, preferredTimescale: 60)) { [weak self, unowned uiView] time in
> 343       guard let self else { return }
> 344       
> 345       Task { @MainActor in  // <--- !!!CRASH HERE !!!
> 346           self.videoPlayer.model.currentTime = uiView.currentDuration
> 347       }
> 348       
> 349       self.updateBuffer(uiView: uiView)
> 350   }
> 

Oh! I didn't see any of that. I can change the unowned references to weak references. Use of unowned should be rare anyway. Man, this is a lot of work for $26...

jb55 commented:
On Tue, Sep 05, 2023 at 12:00:54PM -0700, Kyle Roucis wrote:
>> >No it functions: it swaps between the available functionality for content fitting: .aspectFill and .aspectFit. There does not seem to be "fullscreen" support with the current video player dependency. Maybe @bryanmontz 's upcoming changes can support more, but there was no clear way to implement "fullscreen", just different fitting ratios.
>>
>> ok we should remove that button then until we have fullscreen.
>>
>
>Oh! Uh are you sure? I still find the aspect change to be helpful even without fullscreen support.

yes, the button did nothing on all the videos I tried it on.

kroucis commented:
> I'm hitting crashes, here one I just hit when testing this branch Fatal error: Attempted to read an unowned reference but the object was already deallocated Thread 1 Queue : com.apple.main-thread (serial) #0 0x00000001d55c401c in __pthread_kill () #1 0x00000001f7ded680 in pthread_kill () #2 0x0000000197ea3bb0 in abort () #3 0x00000001897094d8 in swift::fatalErrorv(unsigned int, char const*, char*) () #4 0x00000001897094f8 in swift::fatalError(unsigned int, char const*, ...) () #5 0x00000001897096b4 in swift::swift_abortRetainUnowned(void const*) () #6 0x0000000189766018 in swift_unknownObjectUnownedLoadStrong () #7 0x00000001006cafdc in closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) at /Users/jb55/dev/github/damus-io/damus/damus/Views/Video/VideoPlayer.swift:345 #8 0x00000001006cfe1c in partial apply for closure #1 in closure #7 in VideoPlayer.Coordinator.startObserver(uiView:) () #9 0x00000001006c7a9c in thunk for @escaping @callee_guaranteed @sendable @async () -> () () #10 0x00000001006cff60 in partial apply for thunk for @escaping @callee_guaranteed @sendable @async () -> () () #11 0x000000010069dc54 in thunk for @escaping @callee_guaranteed @sendable @async () -> ***@***.*** τ_0_0) () #12 0x000000010069ddb0 in partial apply for thunk for @escaping @callee_guaranteed @sendable @async () -> ***@***.*** τ_0_0) () VideoPlayer.swift:342 342 observer = uiView.addPeriodicTimeObserver(forInterval: .init(seconds: 0.25, preferredTimescale: 60)) { [weak self, unowned uiView] time in 343 guard let self else { return } 344 345 Task { @mainactor in // <--- !!!CRASH HERE !!! 346 self.videoPlayer.model.currentTime = uiView.currentDuration 347 } 348 349 self.updateBuffer(uiView: uiView) 350 }

I have removed the fullscreen button and changed the unowned references to weak. I have been unable to recreate the crash you listed, so any more information would be helpful. Also removing the button fails the acceptance criteria for this issue. What is to be done about that?

jb55 commented:
>I have removed the fullscreen button and changed the unowned references to weak. I have been unable to recreate the crash you listed, so any more information would be helpful. Also removing the button fails the acceptance criteria for this issue. What is to be done about that?

Not sure what you mean, isn't the acceptance criteria is to be able to
enter an exit fullscreen? This PR doesn't add fullscreen like you
mentioned?

bryanmontz commented on damus/Views/Video/VideoPlayer.swift:
```diff
@@ -166,27 +188,33 @@ public extension VideoPlayer {
+fileprivate extension AVPlayer {
+    var videoSize: CGSize? {
+        get async {
+            let res = Task.detached(priority: .background) {
```
It's not a good idea to use `.background` priority for something that you want to update the UI for the user. I would recommend using the default value, which is `.userInitiated`.

bryanmontz reviewed (commented):
In my opinion, while this is a sorely needed feature add, this PR adds complexity to an already confusing architecture for the video player. I think the addition of UI controls could be done in a way that doesn't interfere so much with the player view, which ideally should just be rendering the video based on the current video settings. There are also lots of non-functional changes (such as changing `mute_icon_color` to `muteIconColor`) mixed in with functional changes, which makes it more difficult to review.

jb55 commented:
@kroucis thanks for your efforts here, try not to get discouraged from review we're just trying to help. If it's not worth your time anymore than that's fair, let us know if it's ok for others to continue this work!

kroucis commented:
> @kroucis thanks for your efforts here, try not to get discouraged from review we're just trying to help. If it's not worth your time anymore than that's fair, let us know if it's ok for others to continue this work!

Yes, I understand, I've been in software for a long time. But it's clear that this approach to the issue is not acceptable, so another attempt will have to be made. Once @bryanmontz 's video player changes have been made and introduced, a simpler UI that is rendering the video based on the current video settings can be implemented. Since the fullscreen criteria cannot be met at this time, the reviewers have decided these changes are too complex and hard to review, and there may be some unowned reference crash issues, I think it's best to learn from this attempt and try again.

jb55 commented:
On Wed, Sep 06, 2023 at 09:06:21AM -0700, Kyle Roucis wrote:
>> @kroucis thanks for your efforts here, try not to get discouraged from review we're just trying to help. If it's not worth your time anymore than that's fair, let us know if it's ok for others to continue this work!
>
>Yes, I understand, I've been in software for a long time. But it's clear that this approach to the issue is not acceptable, so another attempt will have to be made. Once @bryanmontz 's video player changes have been made and introduced, a simpler UI that is rendering the video based on the current video settings can be implemented. Since the fullscreen criteria cannot be met at this time, the reviewers have decided these changes are too complex and hard to review, and there may be some unowned reference crash issues, I think it's best to learn from this attempt and try again.

ah thanks for clarifying, sounds good and agreed!
